### PR TITLE
sys_usbd: Fix up sys_usbd_get_descriptor() error handling according to hardware test

### PR DESCRIPTION
**Pull request by brian218**
Today I carried out a hardware test for this syscall in response to @elad335's last comment in #14900 yesterday.
I tested the following conditions on my real hardware, and fixed up the error handling of this syscall accordingly:
1. descriptor=NULL, desc_size=0x32 -> CELL_EINVAL
2. descriptor=*ptr, desc_size=0 -> CELL_ENOMEM
3. descriptor=NULL, desc_size=0 -> CELL_EINVAL

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp [hunk 1]
```diff
@@ -854,14 +854,14 @@ error_code sys_usbd_get_descriptor(ppu_thread& ppu, u32 handle, u32 device_handl
 
 	std::lock_guard lock(usbh.mutex);
 
-	if (!usbh.is_init || !usbh.handled_devices.count(device_handle))
+	if (!usbh.is_init || !usbh.handled_devices.count(device_handle) || !descriptor)
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:857]: This single check can be put out of lock scope.

File: rpcs3/Emu/Io/usb_device.cpp [hunk 2]
```diff
@@ -200,75 +200,64 @@ bool usb_device_emulated::open_device()
 
 u32 usb_device_emulated::get_descriptor(u8 type, u8 index, u8* buf, u32 buf_size)
 {
-	u32 expected_count = 2;
+	const std::array<u8, 2> header {header.size(), type};
+	u32 expected_count = buf ? std::min<u32>(header.size(), buf_size) : 0;
+	memcpy(buf, header.data(), expected_count);
 
-	if (buf_size < expected_count)
+	if (expected_count >= header.size())
```
Comment (elad335) [inline rpcs3/Emu/Io/usb_device.cpp:207]: Avoid changing so many lines, use an early return when this evaluates to false.

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Okay. Done.

File: rpcs3/Emu/Io/usb_device.cpp [hunk 3]
```diff
@@ -200,24 +200,20 @@ bool usb_device_emulated::open_device()
 
 u32 usb_device_emulated::get_descriptor(u8 type, u8 index, u8* buf, u32 buf_size)
 {
-	u32 expected_count = 2;
+	const std::array<u8, 2> header {header.size(), type};
```
Comment (elad335) [inline rpcs3/Emu/Io/usb_device.cpp:203]: size() is not static, isn't that a UB or may return unspecified value? Im not sure what the c++ rules are for this.

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp:203]: Since it's std::array instead of the other STL containers like vector or map, its size() is constexpr according to https://en.cppreference.com/w/cpp/container/array/size

Comment (elad335) [reply rpcs3/Emu/Io/usb_device.cpp:203]: Yeah but this rule applies here? not every constexpr function may actually be constexpr at all times, this is what consteval is for. Im asking because Im not sure it's a good idea to call non-statc methods before an object is initialized.

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp:203]: Would this be any better?
```C++
std::array<u8, 2> header;
header = {header.size(), type};
```

Comment (elad335) [reply rpcs3/Emu/Io/usb_device.cpp:203]: Yeah

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Updated.

File: rpcs3/Emu/Io/usb_device.cpp [hunk 4]
```diff
@@ -200,24 +200,20 @@ bool usb_device_emulated::open_device()
 
 u32 usb_device_emulated::get_descriptor(u8 type, u8 index, u8* buf, u32 buf_size)
 {
-	u32 expected_count = 2;
+	std::array<u8, 2> header; header = {header.size(), type}; // Reference: https://github.com/RPCS3/rpcs3/pull/14903#discussion_r1429197999
```
Comment (elad335) [inline rpcs3/Emu/Io/usb_device.cpp:203]: It's done automatically via git history, that's I suggested earlier to not change many lines which are not part of the pr.

Comment (elad335) [reply rpcs3/Emu/Io/usb_device.cpp:203]: Also put a newline inbetween

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Done. Thank you.
